impeller/compiler/spirv_sksl.cc

```diff
@@ -0,0 +1,412 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/compiler/spirv_sksl.h"
+
+using namespace spv;
+using namespace SPIRV_CROSS_NAMESPACE;
+
+namespace impeller {
+namespace compiler {
+
+std::string CompilerSkSL::compile() {
+  ir.fixup_reserved_names();
+
+  if (get_execution_model() != ExecutionModelFragment) {
+    SPIRV_CROSS_THROW("Only fragment shaders are supported.'");
+    return "";
+  }
+
+  options.es = false;
+  options.version = 300;
+  options.vulkan_semantics = false;
+  options.enable_420pack_extension = false;
+
+  backend.allow_precision_qualifiers = false;
+  backend.basic_int16_type = "short";
+  backend.basic_int_type = "int";
+  backend.basic_uint16_type = "ushort";
+  backend.basic_uint_type = "uint";
+  backend.double_literal_suffix = false;
+  backend.float_literal_suffix = false;
+  backend.long_long_literal_suffix = false;
+  backend.needs_row_major_load_workaround = true;
+  backend.nonuniform_qualifier = "";
+  backend.support_precise_qualifier = false;
+  backend.uint32_t_literal_suffix = false;
+  backend.use_array_constructor = true;
+  backend.workgroup_size_is_hidden = true;
+
+  fixup_anonymous_struct_names();
+  fixup_type_alias();
+  reorder_type_alias();
+  build_function_control_flow_graphs_and_analyze();
+  fixup_image_load_store_access();
+  update_active_builtins();
+  analyze_image_and_sampler_usage();
+  analyze_interlocked_resource_usage();
+
+  uint32_t pass_count = 0;
+  do {
+    reset(pass_count);
+
+    // Move constructor for this type is broken on GCC 4.9 ...
+    buffer.reset();
+
+    emit_header();
+    emit_resources();
+
+    emit_function(get<SPIRFunction>(ir.default_entry_point), Bitset());
+
+    pass_count++;
+  } while (is_forcing_recompilation());
+
+  statement("half4 main(float2 ignored)");
+  begin_scope();
+  statement("  __main();");
+  statement("  return " + output_name_ + ";");
+  end_scope();
+
+  return buffer.str();
+}
+
+void CompilerSkSL::emit_header() {
+  statement("// This SkSL shader is autogenerated by spirv-cross.");
+  statement("");
+
+  // This builtin is copied from
```

Comment:
The near-term goal is to mimic the SPIR-V -> SkSL [transpiler](https://github.com/flutter/engine/tree/main/lib/spirv/lib/src) in the Engine so that we can remove it. The main constraint is that we can't break the [ink_sparkle shader](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/shaders/ink_sparkle.frag). The ink_sparkle shader references `gl_FragCoord`, and the existing transpiler translates that to `float4(iFragCoord, 0, 0)`, where `iFragCoord` is the input of SkSL's `main()`. I've modified this PR to do the same thing instead of hacking in `sk_FragCoord`.

It sounds like going forward we'll have to think a bit harder about this setup if we're going to aim for parity between the different backends.